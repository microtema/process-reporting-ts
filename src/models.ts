export enum ReportStatus {
    /**
     * Process/Activity has been queued
     */
    QUEUED = 'QUEUED',
    /**
     * Process/Activity has been started
     */
    STARTED = 'STARTED',
    /**
     * Process/Activity has been restarted
     */
    RESTARTED = 'RESTARTED',
    /**
     * Process/Activity has been completed
     */
    COMPLETED = 'COMPLETED',
    /**
     * Process/Activity has been terminated and should be deleted
     */
    TERMINATED = 'TERMINATED',

    PROCESS_COMPLETED = 'PROCESS_COMPLETED',
    /**
     * Process/Activity has warning, but will not stop the process
     */
    WARNING = 'WARNING',
    /**
     * Process/Activity has error and the process will be terminated
     */
    ERROR = 'ERROR',
}

export interface ReportEvent {
    /**
     * Camunda Element id
     */
    elementId: string;

    transactionId: string;

    /**
     * Camunda process id
     */
    processId: string;

    /**
     * Unique Execution id
     */
    executionId: string;

    /**
     * Camunda process version 1.0
     */
    processVersion: string;

    /**
     * Error message
     */
    errorMessage?: string;

    /**
     * Event status, [STARTED, COMPLETED, ERROR, ...]
     */
    status: ReportStatus;

    /**
     * Event timestamp
     */
    eventTime: Date;

    /**
     * Technical meta information of retry the same event
     */
    retryCount?: number;

    /**
     * NOTE: if size of payload or error message is longer then supported, we need to chunked in multiple parts
     */
    partCount: number;

    /**
     * Reference id/key of this event
     */
    referenceId?: string;

    /**
     * Bounded Context of this event, [user, invoice, billing, ...]
     */
    referenceType?: string;

    /**
     * Who trigger the event, [user, technical-user, machine, ...]
     */
    startedBy: string;

    /**
     * Multiple Instance Count for the same activity butt different data
     */
    multipleInstanceIndex: string;

    /**
     * Serialized JSON payload
     */
    payload?: string;

    /**
     * Event meta information
     */
    params: any;

    startEvent: boolean;

    endEvent: boolean;
}

export interface BpmnElementOptions {
    maskParams?: Record<string, string | Function>
    id?: string
    processId?: string
    multipleInstance?: string
    retries?: number
    delay?: number
    suppressException?: boolean
    startEvent?: boolean
    endEvent?: boolean
    instanceIdExpression?: string | Function
    keyExpression?: string | Function
    startedByExpression?: string | Function
    transactionIdExpression?: string | Function
}

export interface ProcessReportingConfiguration {
    disabled: boolean
    processId: string
    processName: string
    processDescription: string
    retentionTime: string
    processVersion: string
    server: string
    maxEventContentLimit: number //2000;
    startElementId: string
}

export interface EventData {
    /**
     * Unique ID of event, UUID
     */
    eventId: string

    /**
     * Who trigger the event, [user, technical-user, machine, ...]
     */
    eventTriggeredBy: string

    /**
     * Category of event, [user, invoice, billing, ...]
     */
    eventCategory: string

    /**
     * Version of Event
     */
    eventVersion: string

    /**
     * Timestamp of event
     */
    eventTriggeredAt: Date

    /**
     * Name of event, [UserCreated, UserDeleted, UserUpdated, ...]
     */
    eventName: string

    /**
     * JSON as String to be converted to generic type, or any generic type that is serializable
     */
    eventAttachment: any

    /**
     * JSON as String to be converted to generic type, or any generic type that is serializable
     */
    eventPayload: any

    /**
     * Origin of event, [application-a, system-b, component-c]
     */
    eventOrigin: string

    /**
     * Technical meta information of retry the same event
     */
    eventRetryCount: number

    /**
     * Unique execution id of event, same event id may have different execution id
     */
    eventExecutionId: string
}