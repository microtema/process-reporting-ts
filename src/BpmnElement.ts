import {v4 as uuidv4} from 'uuid';

import {
    ProcessReportingService,
    ReportEvent,
    ReportStatus,
    getParameters,
    waitFor,
    getReferenceId,
    BpmnElementOptions,
    EventData
} from './index';

const processReportingService = new ProcessReportingService()

export function BpmnElement(options: BpmnElementOptions): DecoratorFunction {

    let transactionId:string | undefined;
    if (options.startEvent) {
       transactionId = uuidv4()
    }

    /**
     * target: The prototype of the class (Object)
     * propertyKey: The name of the method (string | symbol).
     * descriptor: A TypedPropertyDescriptor â€” see the type, leveraging the Object.defineProperty under the hood.
     *
     * NOTE: It's very important here we do not use arrow function otherwise 'this' will be messed up due
     * to the nature how arrow function defines this inside.
     *
     */
    return function (target: Record<string, any>, propertyKey: string, descriptor: TypedPropertyDescriptor<any>) {

        const originalFunc = descriptor.value;

        descriptor.value = function async(...args: any[]) {

            //  console.log(`Start ${propertyKey} function`, `${this.constructor.name}.${propertyKey}`);

            const params = getParameters(originalFunc)

            const payload = getPayload(params, args, options.startEvent === true)
            const referenceId = getReferenceId(payload, options.keyExpression)

            const reportEvent = {
                payload: JSON.stringify(payload),
                referenceId: referenceId,
                status: ReportStatus.STARTED,
                transactionId: transactionId,
                eventTime: new Date()
            } as ReportEvent

            return descriptorFunction(originalFunc, this, args, options, reportEvent, 0)

            // filter is parameter names: to not log them

            //  logger.debug(`${propertyKey} function Input Names is ${params}`);
            //  logger.debug(`${propertyKey} function Input Parameters is ${args}`);
        };

        return descriptor;
    }
}

function publishEvent(options: BpmnElementOptions, reportEvent: ReportEvent) {

    reportEvent.elementId = options.bpmnElementId

    if (options.processId) {
        reportEvent.processId = options.processId
    }

    reportEvent.startEvent = options.startEvent === true
    reportEvent.endEvent = options.endEvent === true

    processReportingService.fireEvent(reportEvent)
}

async function descriptorFunction(originalFunc: Function, scope: any, args: any, options: BpmnElementOptions, reportEvent: ReportEvent, retryCount: number) {

    publishEvent(options, reportEvent)

    try {

        // console.log('args', args)
        const answer = originalFunc.apply(scope, args)
        // console.log('answer', answer)

        const payload = JSON.stringify(answer)
        const status = ReportStatus.COMPLETED

        const newReportEvent = {...reportEvent, ...{payload, status}}

        publishEvent(options, newReportEvent);

        return answer;
    } catch (e: any) {

        const retries = Math.max(0, options.retries || 0)
        const delay = Math.max(250, options.delay || 0)

        const newReportEvent = {...reportEvent, ...{errorMessage: e}}

        if (options.suppressException) {

            newReportEvent.status = ReportStatus.WARNING

            publishEvent(options, newReportEvent)
            return null
        }

        if (retryCount == retries) {

            newReportEvent.status = ReportStatus.ERROR
            publishEvent(options, newReportEvent)
            return null
        }

        await waitFor(delay);

        newReportEvent.status = ReportStatus.RESTARTED

        return descriptorFunction(originalFunc, scope, args, options, newReportEvent, retryCount + 1)
    }
}

export function getPayload(parameterNames: string[], methodArguments: any[], startEvent: boolean) {

    const payload: any = {}

    for (let index = 0; index < parameterNames.length; index++) {

        const param = parameterNames[index];
        let methodArgument = methodArguments[index];

        if (startEvent) {

            // This is entry point on Controller
            const eventData = eventDataConverter(methodArgument);

            // enrichEventData(eventData);

            // ReportingLocalContextHolder.set(eventData);
            // ReportingContextHolder.set(eventData);

            // return eventData.eventAttachment || eventData.eventPayload

        } else if (isByteArray(methodArgument)) {

            // methodArgument = "Binary [ " + FileUtils.byteCountToDisplaySize(length) + " ]";
            methodArgument = "Binary [ " + methodArgument.byteLength + " ]";
        }

        payload[param] = methodArgument
    }

    return payload;
}

function isByteArray(array: any) {

    return array && array.byteLength !== undefined;
}

function eventDataConverter(methodArguments: any[]): EventData {

    return {
        eventAttachment: methodArguments
    } as EventData
}

export type DecoratorFunction = (target: Record<string, any>, propertyKey: string, descriptor: TypedPropertyDescriptor<any>) => TypedPropertyDescriptor<any>;