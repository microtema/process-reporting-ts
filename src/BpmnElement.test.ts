import {BpmnElement} from './BpmnElement'
import {ReportEvent, ReportStatus} from './models'
import axios from 'axios'

jest.mock('axios')
const mockedAxios = axios as jest.Mocked<typeof axios>


describe('Intercept on class method', () => {

    const events: ReportEvent[] = []

    beforeEach(() => {

        events.length = 0

        mockedAxios.post.mockImplementation((url: string, data: any) => {
            events.push(data)
            return Promise.resolve(data)
        })

        process.env.REPORTING_SERVER = 'http://localhost:8080'
        process.env.REPORTING_PROCESS_ID = 'process-id'
        process.env.REPORTING_PROCESS_VERSION = '1.0'
        process.env.REPORTING_EVENT_CATEGORY = 'UnitTest'
    })

    afterEach(() => {

        jest.resetAllMocks();
    });

    test('should be COMPLETED', async () => {

        // Given
        class Service {

            @BpmnElement({
                id: 'StartEvent_1',
                startEvent: true,
                instanceIdExpression: '{{ input }}',
                keyExpression: '{{ input }}'
            })
            execute(input: string): Record<string, string> {
                return {output: input}
            }
        }

        const sut = new Service();
        const input = 'foo'

        // When
        const answer = await sut.execute(input)

        expect(answer).toEqual({output: input})

        // Then
        expect(events.length).toBe(2);
        expect(mockedAxios.post).toHaveBeenCalledWith('http://localhost:8080/reporting-service/rest/api/report', events[0])
        expect(mockedAxios.post).toHaveBeenCalledWith('http://localhost:8080/reporting-service/rest/api/report', events[1])

        expect(events[0].status).toEqual(ReportStatus.STARTED)
        expect(events[0].payload).toEqual(JSON.stringify({input}))
        expect(events[0].elementId).toEqual('StartEvent_1')
        expect(events[0].transactionId).toEqual(input)
        expect(events[0].referenceType).toEqual(process.env.REPORTING_EVENT_CATEGORY)
        expect(events[0].referenceId).toEqual(input)
        expect(events[0].startEvent).toEqual(true)
        expect(events[0].endEvent).toEqual(false)
        expect(events[0].multipleInstanceIndex).toEqual('0.0')
        expect(events[0].errorMessage).not.toBeDefined()
        expect(events[0].processId).toEqual(process.env.REPORTING_PROCESS_ID)
        expect(events[0].processVersion).toEqual(process.env.REPORTING_PROCESS_VERSION)
        expect(events[0].executionId).toBeDefined()
        expect(events[0].eventTime).toBeDefined()
        expect(events[0].startedBy).toEqual('system')


        expect(events[1].status).toEqual(ReportStatus.COMPLETED)
        expect(events[1].payload).toEqual(JSON.stringify({output: input}))
        expect(events[1].elementId).toEqual('StartEvent_1')
        expect(events[1].transactionId).toEqual(input)
        expect(events[1].referenceType).toEqual('UnitTest')
        expect(events[1].referenceId).toEqual(input)
        expect(events[1].startEvent).toEqual(true)
        expect(events[1].endEvent).toEqual(false)
        expect(events[1].multipleInstanceIndex).toEqual('0.0')
        expect(events[1].errorMessage).not.toBeDefined()
        expect(events[1].processId).toEqual(process.env.REPORTING_PROCESS_ID)
        expect(events[1].processVersion).toEqual(process.env.REPORTING_PROCESS_VERSION)
        expect(events[1].executionId).toBeDefined()
        expect(events[1].eventTime).toBeDefined()
        expect(events[1].startedBy).toEqual('system')
    })

    test('should be ERROR', async () => {

        // Given
        class Service {

            @BpmnElement({
                id: 'StartEvent_1',
                endEvent: true,
                instanceIdExpression: '{{ message }}',
                startedByExpression: '{{ message }}',
                keyExpression: '{{ message }}'
            })
            execute(message: string): string {
                throw new Error('Business Error')
            }
        }

        const sut = new Service();
        const message = 'foo'

        // When
        try {
            await sut.execute('foo')
        } catch (e) {
            expect(e).toEqual(new Error('Business Error'))
        }

        // Then
        expect(events.length).toEqual(2)
        expect(mockedAxios.post).toHaveBeenCalledWith('http://localhost:8080/reporting-service/rest/api/report', events[0])
        expect(mockedAxios.post).toHaveBeenCalledWith('http://localhost:8080/reporting-service/rest/api/report', events[1])

        expect(events[0].status).toEqual(ReportStatus.STARTED)
        expect(events[0].payload).toEqual(JSON.stringify({message}))
        expect(events[0].elementId).toEqual('StartEvent_1')
        expect(events[0].transactionId).toEqual(message)

        expect(events[0].referenceId).toEqual(message)
        expect(events[0].startEvent).toEqual(false)
        expect(events[0].endEvent).toEqual(true)
        expect(events[0].multipleInstanceIndex).toEqual('0.0')
        expect(events[0].errorMessage).not.toBeDefined()
        expect(events[0].processId).toEqual(process.env.REPORTING_PROCESS_ID)
        expect(events[0].processVersion).toEqual(process.env.REPORTING_PROCESS_VERSION)
        expect(events[0].executionId).toBeDefined()
        expect(events[0].eventTime).toBeDefined()
        expect(events[0].startedBy).toEqual('foo')

        expect(events[1].status).toEqual(ReportStatus.ERROR)
        expect(events[1].payload).not.toBeDefined()
        expect(events[1].elementId).toEqual('StartEvent_1')
        expect(events[1].transactionId).toEqual(message)

        expect(events[1].referenceId).toEqual(message)
        expect(events[1].startEvent).toEqual(false)
        expect(events[1].endEvent).toEqual(true)
        expect(events[1].multipleInstanceIndex).toEqual('0.0')
        expect(events[1].errorMessage).toEqual('Error: Business Error')
        expect(events[1].processId).toEqual(process.env.REPORTING_PROCESS_ID)
        expect(events[1].processVersion).toEqual(process.env.REPORTING_PROCESS_VERSION)
        expect(events[1].executionId).toBeDefined()
        expect(events[1].eventTime).toBeDefined()
        expect(events[1].startedBy).toEqual('foo')
    })
});

