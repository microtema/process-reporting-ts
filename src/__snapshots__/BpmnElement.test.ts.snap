// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`intercept should be COMPLETED 1`] = `
{
  "key": "foo",
  "value": "bar",
}
`;

exports[`intercept should be COMPLETED 2`] = `
[
  {
    "elementId": "StartEvent_1",
    "endEvent": false,
    "eventTime": 2023-11-16T18:44:49.350Z,
    "payload": "{"foo":"bar"}",
    "referenceId": null,
    "startEvent": true,
    "status": "STARTED",
    "transactionId": "00f63f14-8885-4540-88a3-01e778091a44",
  },
  {
    "elementId": "StartEvent_1",
    "endEvent": false,
    "eventTime": 2023-11-16T18:44:49.350Z,
    "payload": "{"key":"foo","value":"bar"}",
    "referenceId": null,
    "startEvent": true,
    "status": "COMPLETED",
    "transactionId": "00f63f14-8885-4540-88a3-01e778091a44",
  },
]
`;

exports[`intercept should be COMPLETED with referenceId 1`] = `"event-id"`;

exports[`intercept should be ERROR 1`] = `null`;

exports[`intercept should be ERROR 2`] = `
[
  {
    "elementId": "StartEvent_1",
    "endEvent": false,
    "eventTime": 2023-11-16T18:44:49.352Z,
    "payload": "{"foo":"foo"}",
    "referenceId": null,
    "startEvent": false,
    "status": "STARTED",
    "transactionId": undefined,
  },
  {
    "elementId": "StartEvent_1",
    "endEvent": false,
    "errorMessage": [Error: Unable to execute!],
    "eventTime": 2023-11-16T18:44:49.352Z,
    "payload": "{"foo":"foo"}",
    "referenceId": null,
    "startEvent": false,
    "status": "ERROR",
    "transactionId": undefined,
  },
]
`;

exports[`intercept should be RESTARTED 1`] = `null`;

exports[`intercept should be RESTARTED 2`] = `
[
  {
    "elementId": "StartEvent_1",
    "endEvent": false,
    "eventTime": 2023-11-16T18:44:49.354Z,
    "payload": "{"foo":"foo"}",
    "referenceId": null,
    "startEvent": false,
    "status": "STARTED",
    "transactionId": undefined,
  },
  {
    "elementId": "StartEvent_1",
    "endEvent": false,
    "errorMessage": [Error: Unable to convert!],
    "eventTime": 2023-11-16T18:44:49.354Z,
    "payload": "{"foo":"foo"}",
    "referenceId": null,
    "startEvent": false,
    "status": "RESTARTED",
    "transactionId": undefined,
  },
  {
    "elementId": "StartEvent_1",
    "endEvent": false,
    "errorMessage": [Error: Unable to convert!],
    "eventTime": 2023-11-16T18:44:49.354Z,
    "payload": "{"foo":"foo"}",
    "referenceId": null,
    "startEvent": false,
    "status": "RESTARTED",
    "transactionId": undefined,
  },
  {
    "elementId": "StartEvent_1",
    "endEvent": false,
    "errorMessage": [Error: Unable to convert!],
    "eventTime": 2023-11-16T18:44:49.354Z,
    "payload": "{"foo":"foo"}",
    "referenceId": null,
    "startEvent": false,
    "status": "ERROR",
    "transactionId": undefined,
  },
]
`;

exports[`intercept should be WARNING 1`] = `null`;

exports[`intercept should be WARNING 2`] = `
[
  {
    "elementId": "StartEvent_1",
    "endEvent": false,
    "eventTime": 2023-11-16T18:44:49.353Z,
    "payload": "{"foo":"foo"}",
    "referenceId": null,
    "startEvent": false,
    "status": "STARTED",
    "transactionId": undefined,
  },
  {
    "elementId": "StartEvent_1",
    "endEvent": false,
    "errorMessage": [Error: Unable to convert!],
    "eventTime": 2023-11-16T18:44:49.353Z,
    "payload": "{"foo":"foo"}",
    "referenceId": null,
    "startEvent": false,
    "status": "WARNING",
    "transactionId": undefined,
  },
]
`;
